File: p039m/combination_sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  candidates.sort((a, b) => a - b);
  const ans = [];
  /**
   * Recursive Backtrack helper function
   * closure {number[][]} solutions: output
   * closure {number[]} candidates
   * @param {number[]} workPath:  temporary working path
   * @param {number}   remTarget: ramaining target
   * @param {number}   start: index in candidates to start search
   */
  const recBacktrack = function (workPath, remTarget, start) {
    if (remTarget < 0) return;
    if (remTarget === 0) {
      ans.push(Array.from(workPath));
      return;
    }
    for (let i = start; i < candidates.length; ++i) {
      workPath.push(candidates[i]);
      recBacktrack(workPath, remTarget - candidates[i], i);
      workPath.pop();
    }
  };
  recBacktrack([], target, 0);
  return ans;
};
[
  [[2, 3, 6, 7], 7],
  [[3, 4, 5, 6, 7, 8, 9, 11, 12], 15],
  [[1, 2, 3], 4],
].forEach(([candidates, target]) => {
  console.log(
    'Combinations in',
    candidates,
    'sum to',
    target,
    '->',
    combinationSum(candidates, target)
  );
});

File: p039m/combination_sum.py
class Solution:
    def combinationSum(
        self, candidates: list[int], target: int
    ) -> list[list[int]]:
        candidates.sort()
        ans = []

        def helper(path: list[int], target: int, start: int) -> None:
            if target < 0:
                return
            if target == 0:
                ans.append(list(path))
                return
            for i in range(start, len(candidates)):
                path.append(candidates[i])
                helper(path, target - candidates[i], i)
                path.pop()

        helper([], target, 0)
        return ans


# TESTS
for candidates, target, expected in [
    ([2, 3, 6, 7], 7, [[2, 2, 3], [7]]),
    ([2, 3, 5], 8, [[2, 2, 2, 2], [2, 3, 3], [3, 5]]),
    ([2], 1, []),
    ([1], 1, [[1]]),
    ([1], 2, [[1, 1]]),
]:
    sol = Solution()
    actual = sol.combinationSum(candidates, target)
    print("Combinations in", candidates, "sum to", target, "->", actual)
    assert actual == expected
